apirest.md

```diff
@@ -1220,6 +1224,335 @@ $ curl -X DELETE \
+  }
+]
+```
+
+## Get available massive actions for an item
+
+* **URL**: apirest.php/getMassiveActions/:itemtype/:id
+* **Description**: Show the availables massive actions for a given item.
+* **Method**: GET
+* **Parameters**: (Headers)
+  * *Session-Token*: session var provided by [initSession](#init-session) endpoint. Mandatory.
+  * *App-Token*: authorization string provided by the GLPI API configuration. Optional.
+* **Parameters**: (query string)
+* **Returns**:
+  * 200 (OK).
+  * 400 (Bad Request) with a message indicating an error in input parameter.
+  * 401 (UNAUTHORIZED).
+
+Example usage (CURL):
+
+```bash
+$ curl -X GET \
+-H 'Content-Type: application/json' \
+-H "Session-Token: 83af7e620c83a50a18d3eac2f6ed05a3ca0bea62" \
+-H "App-Token: f7g3csp8mgatg5ebc5elnazakw20i9fyev1qopya7" \
+'http://path/to/glpi/apirest.php/getMassiveActions/Computer/3'
+
+< 200 OK
+[
+  {
+    "key": "MassiveAction:update",
+    "label": "Update"
+  },
+  {
+    "key": "MassiveAction:clone",
+    "label": "Clone"
+  },
+  {
+    "key": "Infocom:activate",
+    "label": "Enable the financial and administrative information"
+  },
+  {
+    "key": "MassiveAction:delete",
+    "label": "Put in trashbin"
+  },
+  {
+    "key": "MassiveAction:add_transfer_list",
+    "label": "Add to transfer list"
+  },
+  {
+    "key": "Appliance:add_item",
+    "label": "Associate to an appliance"
+  },
+  {
+    "key": "Item_OperatingSystem:update",
+    "label": "Operating systems"
+  },
+  {
+    "key": "Computer_Item:add",
+    "label": "Connect"
+  },
+  {
+    "key": "Item_SoftwareVersion:add",
+    "label": "Install"
+  },
+  {
+    "key": "KnowbaseItem_Item:add",
+    "label": "Link knowledgebase article"
+  },
+  {
+    "key": "Document_Item:add",
+    "label": "Add a document"
+  },
+  {
+    "key": "Document_Item:remove",
+    "label": "Remove a document"
+  },
+  {
+    "key": "Contract_Item:add",
+    "label": "Add a contract"
+  },
+  {
+    "key": "Contract_Item:remove",
+    "label": "Remove a contract"
+  },
+  {
+    "key": "MassiveAction:amend_comment",
+    "label": "Amend comment"
+  },
+  {
+    "key": "MassiveAction:add_note",
+    "label": "Add note"
+  },
+  {
+    "key": "Lock:unlock",
+    "label": "Unlock components"
+  }
+]
+
+
+$ curl -X GET \
+-H 'Content-Type: application/json' \
+-H "Session-Token: 83af7e620c83a50a18d3eac2f6ed05a3ca0bea62" \
+-H "App-Token: f7g3csp8mgatg5ebc5elnazakw20i9fyev1qopya7" \
+'http://path/to/glpi/apirest.php/getMassiveActions/Computer/4'
+
+< 200 OK
+[
+  {
+    "key": "MassiveAction:purge_item_but_devices",
+    "label": "Delete permanently but keep devices"
+  },
+  {
+    "key": "MassiveAction:purge",
+    "label": "Delete permanently and remove devices"
+  },
+  {
+    "key": "MassiveAction:restore",
+    "label": "Restore"
+  },
+  {
+    "key": "Lock:unlock",
+    "label": "Unlock components"
+  }
+]
+```
+
+## Get massive action parameters
+
+* **URL**: apirest.php/getMassiveActionParameters/:itemtype/
```

Comment:
Some massive actions parameters may depend on given input, see `Change_Ticket:solveticket`, `Group:changegroup`, `Item_SoftwareLicense:move_license`, `Item_SoftwareVersion:move_version`, `Problem_Ticket:solveticket`, `Rule:move_rule`, `Ticket:merge_as_followup` and maybe other cases (cases I missed in GLPI core, or specific cases in plugins).

Maybe `ids` parameter should be required in this method.

Reply:
I don't really know if we can improve that...

`getMassiveActionParameters` feels like a mandatory endpoint as the user need to know what are the expected parameters but on the other hand the code handling massive action does not have any way to exposes these parameters properly (and is IMO a very ugly piece of code which would really need a clean rework 🙄 and was definitely never expected to work through the API).

Looking at some of your example, seems like there is a lot of extra 'unexpected' parameters, for example:
`Group:changegroup` requires `$input['is_tech']`, `$input['check_items_id']` and `$input['check_itemtype']` ?
I don't really see how to handle cases like this where the code that handle the parameters depends itself on some others unpredictable parameters ?


IMO the current code that parse the forms to guess the inputs is feels like a hacky band aid and this endpoint can't fully work without breaking compatibility with existing massive actions (by developing some kind of interface each massive actions must implements, with a part of it dedicated to exposing the actions parameters' properly).

Maybe add a warning in the doc that this endpoint is experimental and incomplete until some other work on massive actions can be done ?

Reply:
> Maybe add a warning in the doc that this endpoint is experimental and incomplete until some other work on massive actions can be done ?

Adding a warning is indeed a good idea, as it seems difficult to handle these cases for now.